validation.go

```diff
@@ -0,0 +1,143 @@
+package merkle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+const MaxUint = ^uint(0)
+
+// ValidatePartialTree uses leafIndices, leaves and proof to calculate the merkle root of the tree and then compares it
+// to expectedRoot.
+func ValidatePartialTree(leafIndices []uint64, leaves, proof [][]byte, expectedRoot []byte,
+	hash func(lChild, rChild []byte) []byte) (bool, error) {
+	v, err := newValidator(leafIndices, leaves, proof, hash)
+	if err != nil {
+		return false, err
+	}
+	root, err := v.calcRoot(MaxUint)
+	return bytes.Equal(root, expectedRoot), err
+}
+
+func newValidator(leafIndices []uint64, leaves, proof [][]byte,
+	hash func(lChild, rChild []byte) []byte) (validator, error) {
+	if len(leafIndices) != len(leaves) {
+		return validator{}, fmt.Errorf("number of leaves (%d) must equal number of indices (%d)", len(leaves),
+			len(leafIndices))
+	}
+	if len(leaves) == 0 {
+		return validator{}, fmt.Errorf("at least one leaf is required for validation")
+	}
+	proofNodes := &proofIterator{proof}
+	leafIt := &leafIterator{leafIndices, leaves}
+
+	return validator{leaves: leafIt, proofNodes: proofNodes, hash: hash}, nil
+}
+
+type validator struct {
+	leaves     *leafIterator
+	proofNodes *proofIterator
+	hash       func(lChild, rChild []byte) []byte
+}
+
+func (v *validator) calcRoot(stopAtLayer uint) ([]byte, error) {
+	idx, activeNode, err := v.leaves.next()
+	if err != nil {
+		return nil, err
+	}
+	p := position{index: idx}
```

Comment:
consider renaming to "activeP"

Reply:
In general, it would good if you can name variables in a way that the reader can tell if the var represents a value of a node or a position, it will improve readability

Reply:
I've renamed everything that's a position to `somethingPos` (there's now `activePos` and `nextLeafPos`), everything else is the value of a node, so I hope that eliminates the confusion.